### PR TITLE
Scheduler and WebSockets DBus signals

### PR DESCRIPTION
**Pull request by aaannz**
- add support for emitting DBus signals into OpenQA::IPC
```perl
OpenQA::IPC->ipc->emit_signal($source_object, $signal_name, [$options])
```
- add Scheduler signals
job_new, [['array', 'uint32']]
job_duplicate, ['uint32', 'uint32']
job_grab, [['dict', 'string', ['variant']]]
job_finish, ['uint32', 'uint32', 'string']
job_cancel, [['array', 'uint32']]
job_restart, [['array', 'uint32']]
iso_schedule, [['dict', 'string', ['variant']]]
asset_register, [['dict', 'string', ['variant']]]
asset_delete, [['dict', 'string', ['variant']]]

- add WebSockets signals
worker_connected', [['dict', 'string', ['variant']]]
worker_disconnected', ['uint32']
command_sent',        ['uint32', ['variant']]
command_failed',      ['uint32', ['variant']]

DBus policy file is not yet updated in this PR, so to listen to these signals one need to be geekotest. But I plan to open them to other users and also to open some scheduler getters.
Next step is to add signals to WebAPI so we can register template, job groups, machines changes and etc.

### CODE REVIEW

**Comment by coolo** [comment]
I miss the context - what is the use case of this?

**Comment by aaannz** [comment]
Trivial use case is auditing without the need to implement audit support to openQA. By trivial system bus observation one can see what's going on.
When the WebAPI part is done then this ( https://progress.opensuse.org/issues/7478 ) can be done in external app without DB changes.
And then there is Richard's use case of sending emails when some tests are finished (yeah, the test-invoked events are not in, yet).

TLDR; without external tools this is useless, with external tools can be priceless.

**Comment by aaannz** [comment]
I assumed that everyone know what DBus signals are, so for those not familiar with the concept this PR makes openQA to push (broadcast) DBus messages to systembus when some events occur.
So i.e. when new iso is scheduled, openQA broadcasts iso_schedule message with parameters like DISTRI, BUILD, ARCH, etc.. and similar when worker comes online, job is assigned, finishes, etc.

Dbus daemon has a filters so when no one is listening to this signals, none will be emitted.
To listen to this signals (and called methods) simply run:
```bash
dbus-monitor --system "sender=org.opensuse.openqa.WebSockets" "sender=org.opensuse.openqa.Scheduler"
```

**Comment by aaannz** [comment]
@lnussel what do you think about this?

**Comment by lnussel** [comment]
well, fine with me in general. I still think the separate Scheduler process needs to die though.
Mojo also supports events internally http://mojolicio.us/perldoc/Mojo/EventEmitter
so a "native" plugin interface would probably be based on that. Would be in-process of course though.

File: lib/OpenQA/Scheduler/Scheduler.pm [hunk 1]
```diff
@@ -581,6 +583,7 @@ sub job_grab {
 
     # TODO: cleanup previous tmpdir
     $worker->set_property('WORKER_TMPDIR', tempdir());
+    OpenQA::IPC->ipc->emit_signal('scheduler', 'job_grab', $job_hashref);
```
Comment (lnussel) [inline lib/OpenQA/Scheduler/Scheduler.pm]: this would also emit secret properties such as jobtoken, wouldn't it? Maybe it should just emit the job id instead

**Comment by lnussel** [comment]
Then there's also http://mojolicio.us/perldoc/Mojolicious/Guides/Cookbook#EventSource-web-service which would allow to get events via http

File: lib/OpenQA/Scheduler/Scheduler.pm (hunk 1 above)
Comment (aaannz) [inline lib/OpenQA/Scheduler/Scheduler.pm]: Yes, but I was thinking we can limit access to sensitive signals in dbus policy file.
My idea was to add as much info to signals as I can to limit round trips. (get id -> ask details).

But in this case, one probably does not need info how to connect to WebUI so I'll mask it.

**Comment by AdamWill** [comment]
So hopefully not to derail this, but we (Fedora) would like our openQA to emit [fedmsg](http://www.fedmsg.com) signals, and I'm struggling to figure out what would be the best way to implement this so it can go upstream cleanly and be optional.

Could we possibly have an emitter system which is pluggable / modular in some sense - so when the code hits points where it wants to emit signals, it runs all the emitters that are available / configured? I don't know enough about perl/mojo to know if that's reasonable or how best to design it, though.

We *can* write some kind of silly little relay daemon thing which listens out for dbus signals from openQA and relays them to fedmsg, but to me at least it seems like it'd be nicer for it to be in openQA itself, so we don't have another little thing to remember to document and keep working and make sure is always included in deployments and running...

**Comment by lnussel** [comment]
fedmsg sounds interesting indeed

**Comment by AdamWill** [comment]
If folks are willing to give me enough hints about what an upstreamable design for this would be I'm happy to try and hack it up (I like learning processes!), but right now I'm at that point where I just don't know the right way to get started and don't want to spend a lot of time writing something and then find out that it's completely stupid and I should've done it this other way instead :)

**Comment by aaannz** [comment]
Mojo itself has notion of events so one idea is to emit mojo signals instead of dbus signals at all these points (and more) I currently propose in PR. Then 'native' mojo plugins can be written which will forward them to other places (dbus, fedmsg, zmq, ..) or do whatever.
The problem here to solve is that current scheduler is not using mojo message loop.
So either
* reintegrate scheduler back to webapi. This will require to add some DB locking to ensure proper scheduler working.
* or use mojo::ioloop instead of dbus message loop (this is fairly easy to do), but then scheduler would remain standalone (for which I'm still yet to deliver proper use case :) ).

before any mojo emitting.

**Comment by aaannz** [comment]
Yay! Huge PR once more. I left the original history.
This does the same as original implementation but using DBus as a plugin. Events are emitted using Mojo::IOLoop->emit, DBus plugin is subscribed to it and send DBus signal. Apart from IPC usage the plugin can serve as a demo implementation.

This PR does not yet contain automatic nor config based plugin loading, but I plan to add it. So the target idea is that one can package its own openqa plugin (for WebAPI, WebSockets and Scheduler) and enable it in openQA config file. There is no specific ABI for plugins apart of emitted signals. Plugins has direct access to openqa internals.

**Comment by AdamWill** [comment]
Sounds awesome! Thanks very much. /me desperately tries to clear space in his schedule for openQA hacking

**Comment by aaannz** [comment]
rebased and added c49acbf which loads plugins as specified in openqa.ini

**Comment by aaannz** [comment]
I'll redo this in few new PR's - events, extension support and dbus extension - to make review easier.

I decided to stop calling it plugins, but rather extensions - in my mind for plugins one has defined plugin interface, here we have no such thing except common registration entry point, so I chose to call it extensions :)

**Comment by aaannz** [comment]
Let me close this as all the ideas from this one will be in different PR's.
